internal/upf/upf.go

```diff
@@ -258,3 +275,27 @@ func (u *UPF) collectCollectionTrackingGarbage(ctx context.Context) {
 		expiredKeys = expiredKeys[:0]
 	}
 }
+
+func (u *UPF) listenForTrafficNotifications() {
+	var record ringbuf.Record
+	var event ebpf.DataNotification
+	for {
+		err := u.notificationReader.ReadInto(&record)
+		if errors.Is(err, os.ErrClosed) {
+			return
+		}
+		if err = binary.Read(bytes.NewBuffer(record.RawSample), binary.NativeEndian, &event); err != nil {
+			logger.UpfLog.Error("Failed to decode data notification", zap.Error(err))
+		}
+		logger.UpfLog.Debug("Received notification for", zap.Uint64("SEID", event.LocalSEID), zap.Uint16("PDRID", event.PdrID), zap.Uint8("QFI", event.QFI))
+		if !u.bpfObjects.IsAlreadyNotified(event) {
+			u.bpfObjects.MarkNotified(event)
+			logger.UpfLog.Debug("Notifying SMF of downlink data", zap.Uint64("SEID", event.LocalSEID), zap.Uint16("PDRID", event.PdrID), zap.Uint8("QFI", event.QFI))
+			err = core.SendPfcpSessionReportRequest(context.TODO(), event.LocalSEID, event.PdrID, event.QFI)
+			if err != nil {
+				logger.UpfLog.Warn("Failed to send downlink data notification", zap.Error(err))
+				u.bpfObjects.ClearNotified(event.LocalSEID, event.PdrID, event.QFI)
```

Comment:
Why not MarkNotified right after notifying without errors. Now we're marking as notified prior to notification, notifying, checking if there's an error, then reverting if there was any.

Reply:
That's a good point, I'll rework this.